From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Karol Szwaj <karol.szwaj@gmail.com>
Date: Thu, 13 Mar 2025 11:08:06 +0100
Subject: [PATCH] UPSTREAM: <carry>: Add kube feature gate for global service
 account

TODO: this commit may be dropped once https://github.com/kcp-dev/kcp/pull/3274 and related issues are done

Signed-off-by: Karol Szwaj <karol.szwaj@gmail.com>

On-behalf-of: @SAP karol.szwaj@sap.com
Signed-off-by: Nelo-T. Wallus <n.wallus@sap.com>
---
 pkg/features/kube_features.go       | 3 +++
 pkg/registry/rbac/validation/kcp.go | 5 ++++-
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/pkg/features/kube_features.go b/pkg/features/kube_features.go
--- a/pkg/features/kube_features.go
+++ b/pkg/features/kube_features.go
@@ -2375,6 +2375,9 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 		{Version: version.MustParse("1.32"), Default: false, PreRelease: featuregate.Alpha},
 		{Version: version.MustParse("1.36"), Default: true, PreRelease: featuregate.Beta},
 	},
+	GlobalServiceAccount: {
+		{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Alpha},
+	},
 }
 
 // defaultKubernetesFeatureGateDependencies enumerates the dependencies of any feature gate that
diff --git a/pkg/registry/rbac/validation/kcp.go b/pkg/registry/rbac/validation/kcp.go
--- a/pkg/registry/rbac/validation/kcp.go
+++ b/pkg/registry/rbac/validation/kcp.go
@@ -12,6 +12,8 @@ import (
 	authserviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
 	"k8s.io/apiserver/pkg/authentication/user"
 	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/kubernetes/pkg/features"
 )
 
 const (
@@ -88,6 +90,7 @@ func EffectiveUsers(clusterName logicalcluster.Name, u user.Info) []user.Info {
 
 	var wantAuthenticated bool
 	var wantUnauthenticated bool
+	globalsa := utilfeature.DefaultFeatureGate.Enabled(features.GlobalServiceAccount)
 
 	var recursive func(u user.Info)
 	recursive = func(u user.Info) {
@@ -105,7 +108,7 @@ func EffectiveUsers(clusterName logicalcluster.Name, u user.Info) []user.Info {
 			wantUnauthenticated = wantUnauthenticated || !found
 		}
 
-		if IsServiceAccount(u) {
+		if IsServiceAccount(u) && globalsa {
 			if clusters := u.GetExtra()[authserviceaccount.ClusterNameKey]; len(clusters) == 1 {
 				nsNameSuffix := strings.TrimPrefix(u.GetName(), "system:serviceaccount:")
 				rewritten := &user.DefaultInfo{
